storage/innobase/xtrabackup/src/file_utils.cc

```diff
@@ -100,3 +147,166 @@ int mkdirp(const char *pathname, int Flags, myf MyFlags) {
+          sizeof(cursor->rel_path));
+
+  cursor->fd =
+      my_open(cursor->abs_path, (read_only) ? O_RDONLY : O_RDWR, MYF(MY_WME));
+
+  if (cursor->fd == -1) {
+    return (false);
+  }
+
+  if (my_fstat(cursor->fd, &cursor->statinfo)) {
+
+    datafile_close(cursor);
+
+    return (false);
+  }
+
+  posix_fadvise(cursor->fd, 0, 0, POSIX_FADV_SEQUENTIAL);
+
+  cursor->buf_size = buffer_size;
+  cursor->buf = static_cast<byte *>(
+      my_malloc(PSI_NOT_INSTRUMENTED, cursor->buf_size, MYF(MY_FAE)));
+
+  return (true);
+}
+
+xb_fil_cur_result_t datafile_read(datafile_cur_t *cursor) {
+  ulint to_read;
+  ulint count;
+
+
+  to_read =
+      std::min(cursor->statinfo.st_size - cursor->buf_offset, cursor->buf_size);
+
+  if (to_read == 0) {
+    return (XB_FIL_CUR_EOF);
+  }
+
+  count = my_read(cursor->fd, cursor->buf, to_read, MYF(0));
+  if (count == MY_FILE_ERROR) {
+    return (XB_FIL_CUR_ERROR);
+  }
+
+  posix_fadvise(cursor->fd, cursor->buf_offset, count, POSIX_FADV_DONTNEED);
+
+  cursor->buf_read = count;
+  cursor->buf_offset += count;
+
+  return (XB_FIL_CUR_SUCCESS);
+}
+
+bool restore_sparseness(const char *src_file_path, uint buffer_size,
+                        char error[512]) {
+  datafile_cur_t cursor;
+  size_t page_size = 0;
+  size_t seek = 0;
+  if (!file_has_suffix("ibd", src_file_path)) return true;
+
+  if (!datafile_open(src_file_path, &cursor, false, buffer_size)) {
+    strcpy(error, "Cannot open file");
+    return false;
+  }
+  auto punch_hole_func = [&](const auto page) {
+    if (fil_page_get_type(page) == XB_FIL_PAGE_COMPRESSED ||
+        fil_page_get_type(page) == XB_FIL_PAGE_COMPRESSED_AND_ENCRYPTED) {
+#ifdef UNIV_DEBUG
+      assert(page_size % (size_t)cursor.statinfo.st_blksize == 0);
+#endif
+      size_t compressed_len =
+          ut_calc_align(mach_read_from_2(page + XB_FIL_PAGE_COMPRESS_SIZE_V1) +
+                            XB_FIL_PAGE_DATA,
+                        cursor.statinfo.st_blksize);
+      if (compressed_len < page_size) {
+#ifdef HAVE_FALLOC_PUNCH_HOLE_AND_KEEP_SIZE
+        int ret =
+            fallocate(cursor.fd, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE,
+                      seek + compressed_len, page_size - compressed_len);
+        if (ret != 0) {
+          strcpy(error, "fallocate returned ");
+          std::string err = std::to_string(errno);
+          strcat(error, err.c_str());
+          return false;
+        }
+#endif  // HAVE_FALLOC_PUNCH_HOLE_AND_KEEP_SIZE
+      }
+    }
+    return true;
+  };
+
+  while (datafile_read(&cursor) == XB_FIL_CUR_SUCCESS) {
+    size_t buf_offset = 0;
+    if (cursor.buf_offset == cursor.buf_read) {
+      const uint32_t flags = fsp_header_get_flags(cursor.buf);
+      const ulint ssize = FSP_FLAGS_GET_PAGE_SSIZE(flags);
+      if (ssize == 0) {
+        page_size = XB_UNIV_PAGE_SIZE_ORIG;
+      } else {
+        page_size = ((XB_UNIV_ZIP_SIZE_MIN >> 1) << ssize);
+      }
+    }
+
```

Comment:
Can you please add assertions or checks here on page_size validity?
ie. page_size should fall be min 1k and max 64k page size

Reply:
Done